src/NonceManager.sol

```diff
@@ -0,0 +1,47 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.23;
+
+import {INonceManager} from "./interfaces/INonceManager.sol";
+import {MinimalDelegationStorage, MinimalDelegationStorageLib} from "./libraries/MinimalDelegationStorage.sol";
+
+/// @title NonceManager
+/// @notice A contract that manages nonces to prevent replay attacks
+abstract contract NonceManager is INonceManager {
+    /// @inheritdoc INonceManager
+    function getNonce(uint256 key) public view virtual returns (uint256 nonce);
+
+    /// @inheritdoc INonceManager
+    function invalidateNonce(uint256 nonce) public virtual;
+
+    /// @notice Validates that the provided nonce is valid and increments the sequence number
+    /// @param nonce A 256-bit value where:
+    ///             - Upper 192 bits: the sequence key
+    ///             - Lower 64 bits: must match the expected sequence number for the key
+    /// @dev If valid, increments the sequence number for future nonce validations
+    function _useNonce(uint256 nonce) internal {
+        uint192 key = uint192(nonce >> 64);
+        uint64 seq = uint64(nonce);
+        if (!(MinimalDelegationStorageLib.get().nonceSequenceNumber[key]++ == seq)) {
+            revert InvalidNonce();
+        }
+    }
+
+    /// @notice Increments the sequence for the key in nonce (i.e. upper 192 bits).
+    /// @param nonce A 256-bit value where:
+    ///             - Upper 192 bits: the sequence key
+    ///             - Lower 64 bits: the sequence number to be invalidated
+    /// @dev Can't invalidate >= 2**16 nonces per transaction.
+    function _invalidateNonce(uint256 nonce) internal {
+        uint192 key = uint192(nonce >> 64);
+        uint64 currentSeq = uint64(MinimalDelegationStorageLib.get().nonceSequenceNumber[key]);
+        uint64 targetSeq = uint64(nonce);
+        if (targetSeq < currentSeq) revert InvalidNonce();
+        // Limit the amount of nonces that can be invalidated in one transaction.
+        unchecked {
```

Comment:
I think if you're going for the design where the user has to pass in a target nonce to invalidate up to, we can just do:

```
MinimalDelegationStorageLib.get().nonceSequenceNumber[key] = targetSeq;
```

now that you have the delta check!

Reply:
in this case im invalidating all the nonces up to and including the one thats passed. thats why i set the new nonce to targetSeq + 1